datafusion/core/src/physical_optimizer/enforce_distribution.rs

```diff
@@ -1216,33 +1216,37 @@ fn ensure_distribution(
     .collect::<Result<Vec<_>>>()?;
 
     let children_plans = children.iter().map(|c| c.plan.clone()).collect::<Vec<_>>();
-    plan = if plan.as_any().is::<UnionExec>() && can_interleave(children_plans.iter()) {
-        // Add a special case for [`UnionExec`] since we want to "bubble up"
-        // hash-partitioned data. So instead of
-        //
-        // Agg:
-        //   Repartition (hash):
-        //     Union:
-        //       - Agg:
-        //           Repartition (hash):
-        //             Data
-        //       - Agg:
-        //           Repartition (hash):
-        //             Data
-        //
-        // we can use:
-        //
-        // Agg:
-        //   Interleave:
-        //     - Agg:
-        //         Repartition (hash):
-        //           Data
-        //     - Agg:
-        //         Repartition (hash):
-        //           Data
-        Arc::new(InterleaveExec::try_new(children_plans)?)
+    if let Some(union_exec) = plan.as_any().downcast_ref::<UnionExec>() {
+        if !union_exec.skip_interleave() && can_interleave(children_plans.iter()) {
```

Comment:
I had to remind myself what Interleave did:

https://github.com/apache/arrow-datafusion/blob/63888e853b7b094f2f47f53192a94f38327f5f5a/datafusion/physical-plan/src/union.rs#L286-L317

Do I understand correctly that the problem with switching to InterleaveExec is that it has more than 1 output partition and thus can't be converted to `ProgressiveEval` (due to  this code
 https://github.com/influxdata/influxdb_iox/blob/124f8b481179fd9f6d03f40cfecf40c326ac3127/iox_query/src/physical_optimizer/sort/order_union_sorted_inputs.rs#L153

)

If that is the case, I wonder would it be possible to make ProgressiveEval work with multiple partitions 🤔

Reply:
From the definition Interleave https://github.com/apache/arrow-datafusion/blob/63888e853b7b094f2f47f53192a94f38327f5f5a/datafusion/physical-plan/src/union.rs#L286-L317, data will be grouped by hash-function/hash-partitioning which means we group data and **won't keep them in any order**.  ProgresiveEval needs data to be in order for it to work. 

So if we can solve the problem to have output of Interleave sorted, it will work with ProgressiveEval. However, if I understand correctly, the whole point of Interleave is we do not have to worry and won't keep data sorted. That is why using Interleave in the first place is not what we want here

Reply:
I am bothered by the fact that the flag is on the `UnionExec` because:
1.  that isn't logically a property of the `UnionExec` but rather a behavior we want to change in one of the OptimzerPasses. This seems to couple the plan to the optimizer which I think makes the overall codebase harder to understand
2. There aren't existing examples of this kind of "change the optimizer behavior" flag on plan nodes  (`ConfigOptions` is used for this purpose elsewhere)

Thus, I think we should add a flag to `ConfigOptions` that will affect the behavior of the EnforceDistribution pass, similar to `repartition_file_scans` is passed. This would also give us a good place to document and explain what the option does. 

So I think that instead of checking on some property of the `union_exec` we should use a property of `options` -- aka that is the `ConfigOptions` passed  here:

```rust
fn ensure_distribution(
    dist_context: DistributionContext,
    config: &ConfigOptions,
) -> Result<Transformed<DistributionContext>> {
```

This would be consistent with how other settings are done and would make it clear when this flag got set

The downside is it would potentially affect all `UnionExec`s in the plan, not just specific ones, but I actually think that is easier to understand and will not cause any performance issue

Reply:
@alamb Here is the DF ticket https://github.com/apache/datafusion/issues/10257. Can you have a look. If I understand correctly, I just need to use that new flag to avoid Interleave and I will find a good place to set that config flag. I will create different branch that can be merged directly into DF main branch for this work